src/scripts/interactions.js

```diff
@@ -0,0 +1,216 @@
+    // Handle the end of selection
+    if (interactions.select_end) {
+      interactions.select_end(intersection);
+    }
+  }
+});
+const handleSelect = handlerCommon((intersection) => {
+  const interactions = intersection.object[Interactions];
+  if (interactions) {
+    if (interactions.select) {
+      interactions.select(intersection);
+    }
+  }
+});
+
+export function setupInteractions() {
+  inputSources = XR.session.getInputSources();
+  console.log(inputSources);
+  XR.session.addEventListener('inputsourceschange', handleInputSourcesChange);
+
+  XR.session.addEventListener('select', handleSelect);
+  XR.session.addEventListener('selectstart', handleSelectStart);
+  XR.session.addEventListener('selectend', handleSelectEnd);
+}
+
+const raycaster = new Raycaster();
+function raycast(xrRay) {
+  const { scene } = getCurrentScene();
+
+  const trMatrix = new Matrix4().fromArray(xrRay.matrix);
+
+  // Transformed ray matrix from the current scene matrix world
+  const rMatrix = new Matrix4().multiplyMatrices(scene.matrixWorld, trMatrix);
+
+  raycaster.set(
+    new Vector3().setFromMatrixPosition(rMatrix),
+    new Vector3(0, 0, -1).transformDirection(rMatrix)
+      .normalize()
+  );
+  const intersections = raycaster.intersectObjects(scene.children, true);
+  for (const intersection of intersections) {
+    intersection.point.applyMatrix4(scene.matrixWorld);
+  }
+  return intersections;
+}
+
+const lastObjects = new Map();
+function updateInputSource(inputSource, ray) {
+  // Handle Drag and Drop
+  if (dragAndDrop.has(inputSource)) {
+    const { object, transformMatrix } = dragAndDrop.get(inputSource);
+    const newMatrix = new Matrix4().multiplyMatrices(new Matrix4().fromArray(ray.matrix), transformMatrix);
+    if (object[Interactions].drag) {
+      object[Interactions].drag(newMatrix);
+    } else {
+      object.matrix = newMatrix;
+      object.updateMatrixWorld(true);
+    }
+  }
+
+  for (const intersection of raycast(ray)) {
+    const lastObject = lastObjects.get(inputSource);
+    if (intersection) {
+      if (lastObject !== intersection.object) {
+        if (lastObject && lastObject[Interactions] && lastObject[Interactions].hover_end) {
+          lastObject[Interactions].hover_end();
+        }
+        if (intersection.object[Interactions] && intersection.object[Interactions].hover_start) {
+          intersection.object[Interactions].hover_start(intersection);
+        }
+        lastObjects.set(inputSource, intersection.object);
+      }
+      if (intersection.object[Interactions] && intersection.object[Interactions].hover) {
+        intersection.object[Interactions].hover(intersection);
+      }
+      break; // MAYBE: Handle more than the closest object?
+    } else {
+      if (lastObject) {
+        if (lastObject[Interactions] && lastObject[Interactions].hover_end) {
+          lastObject[Interactions].hover_end();
+        }
+        lastObjects.delete(inputSource);
+      }
+    }
+  }
+}
```

Comment:
This isn't really necessary as it just adds gaze interaction to the magic window. Any selections/hovers should occur from a touch event which the WebXR API handles pretty well itself. The hover event should be called for every frame that the finger is held down on an object while the select event should only occur on the first frame that the finger touches the object.

https://github.com/Group-47-Capstone-2019/Immersive-Web-VR-AR/blob/a7c0fc1e70de8b925e13ac02d1bbb33cf69b7cd0/src/scripts/interactions.js#L178
https://github.com/Group-47-Capstone-2019/Immersive-Web-VR-AR/blob/a7c0fc1e70de8b925e13ac02d1bbb33cf69b7cd0/src/scripts/interactions.js#L186-L197

Reply:
I ended up removing it.  I still think it would be cool (given more time) because at least on my phone I can't get select_start and then later select_end while dragging my finger on the screen.  Instead it just fires a select_end event as soon as my finger moves on he screen.